pbr/util.py

```diff
@@ -9,47 +9,55 @@
 from config import scene_config
 from scene import environment as env
 
-# Import assets from path as defined by asset_list
-# Where asset list ('assets') is a list of two-tuples, each containing
-#   - the dictionary key and
-#   - regex string for each field
-def populate_assets(path, asset_list):
-    # Populate list of assets at path
-    files = os.listdir(path)
-
-    # Create container for asset entries
-    assets = []
-
-    # Initialise field paths as None
-    fields = {}
-    for item in asset_list:
-        fields.update({item[0]: None})
-
-    # Search through each file in folder to try to find raw and mask image paths
-    for file in files:
-        for item in asset_list:
-            result = re.search(item[1], file, re.I)
-            if result is not None:
-                fields.update({item[0]: os.path.join(path, file)})
-
-    # If we have a mandatory field (first field listed in asset_list)
-    if fields[asset_list[0][0]] is not None:
-        assets.append(fields)
-
-    # Populate list of subdirectories at path
-    subdirs = sorted([x for x in files if os.path.isdir(os.path.join(path, x))])
+# Get a list of files and directores in the given path
+# as a tuple with (files, directories)
+def list_files_and_dirs(path, absolute=False):
+    files = []
+    dirs = []
+
+    for child in os.listdir(path):
+        child_path = os.path.join(path, child)
+        if os.path.isdir(child_path):
+            dirs.append(child_path if absolute else child)
+        else:
+            files.append(child_path if absolute else child)
+
+    return (files, dirs)
+
+
+# Populate the given assets list with assets from path as defined by asset_fields
+# Where asset_fields is a list of two-tuples, each containing
+#   - the field's dictionary key and
+#   - the field's regex string
+def populate_assets(assets, path, asset_fields):
+    # Get the files and folders at the given path
+    files, subdirs = list_files_and_dirs(path, absolute=True)
+
+    # Initialise asset with fields as None
+    asset = {}
+    for field in asset_fields:
+        asset.update({field[0]: None})
+
+    # Search through each file and add them to the asset if they match an asset field
+    for file_path in files:
+        for field_name, field_re in asset_fields:
+            if re.search(field_re, file_path, re.I) is not None:
+                asset.update({field_name: file_path})
+
+    # Add the asset if the first field (the mandatory one) has a file
+    if asset[asset_fields[0][0]] is not None:
+        assets.append(asset)
 
     # For each subdirectory, recursively populate assets
-    for subdir in subdirs:
-        assets += populate_assets(os.path.join(path, subdir), asset_list)
+    for subdir in sorted(subdirs):
+        populate_assets(assets, subdir, asset_fields)
```

Comment:
In hindsight, this probably would've been much easier with [pathlib's glob](https://docs.python.org/3/library/pathlib.html#pathlib.Path.glob) but not necessary right now.